scipy/special/stirling2.c

```diff
@@ -0,0 +1,153 @@
+/* c implementation of Stirling numbers of the second kind */
+
+#include <complex.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#ifndef LANCZOS_H
+#include "cephes/lanczos.h"
+#endif
+
+#ifndef MCONF_H
+#include "cephes/mconf.h"
+#endif
+
+// for lambertw function ??
+// #include
+
+/*
+ *
+ *     Stirling numbers of the second kind
+ *
+ *
+ *
+ * SYNOPSIS: Stirling numbers of the second kind count the
+ *  number of ways to make a partition of n distinct elements
+ *  into k non-empty subsets.
+ *
+ * DESCRIPTION: n is the number of distinct elements of the set
+ *  to be partitioned and k is the number of non-empty subsets.
+ *  The values for n < 0 or k < 0 are interpreted as 0. If you
+ *
+ * ACCURACY: The method returns an unsigned integer type and
+ *  the size of the type 32 bits.
+ *
+ * NOTE: this file is *NOT* part of the original ceph distribution and was
+ *  added by Lucas Roberts
+ */
+
+// Dynamic programming approach
+
+double stirling2_dp(int n, int k){
+    if ((n == 0 && k == 0) || (n==1 && k==1)) {
+        return 1.;
+    }
+    if (k <= 0 || k > n || n < 0){
+        return 0.;
+    }
+    int arraySize = k <= n - k + 1 ? k : n - k + 1;
+    double *curr = malloc(arraySize * sizeof(double));
+    for (int i = 0; i < arraySize; i++){
+        curr[i] = 1.;
+    }
+    // TODO: handle overflow correctly, e.g. update to follow
+    // https://github.com/scipy/scipy/pull/18103#discussion_r1132843734
+    // in both outer branches
+    if (k <= n - k + 1) {
+        for (int i = 1; i < n - k + 1; i++){
+            for (int j = 1; j < k; j++){
+                curr[j] = (j + 1) * curr[j] + curr[j - 1];
+                if (curr[j] <= 0){
+                    free(curr);
+                    return -1.; // numeric overflow
+                }
+            }
```

Comment:
This whole rigmarole was only needed because ints don't have infinity. If you're working with and returning doubles, then you can just return infinity here.

Reply:
> This whole rigmarole was only needed because ints don't have infinity. If you're working with and returning doubles, then you can just return infinity if the values get bigger than a double can hold.

Ah yeah, makes sense, will set to infinity.